tree_unique_args/tests/interpreter.rs

```diff
@@ -361,3 +406,153 @@ fn test_interpreter_fib_using_memory() {
     assert_eq!(vm.mem[&(nth as usize)], Value::Num(fib_nth));
     assert!(!vm.mem.contains_key(&(nth as usize + 1)));
 }
+
+impl std::str::FromStr for Expr {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Expr, String> {
+        fn list_expr_to_vec(e: &egglog::ast::Expr) -> Result<Vec<Expr>, String> {
+            if let egglog::ast::Expr::Call(f, xs) = e {
+                match (f.as_str(), xs.as_slice()) {
+                    ("Nil", []) => return Ok(Vec::new()),
+                    ("Cons", [head, tail]) => {
+                        let head = egglog_expr_to_expr(head)?;
+                        let mut tail = list_expr_to_vec(tail)?;
+                        tail.insert(0, head);
+                        return Ok(tail);
+                    }
+                    _ => {}
+                }
+            }
+            Err(String::from("invalid ListExpr"))
+        }
+        fn egglog_expr_to_id(e: &egglog::ast::Expr) -> Result<Id, String> {
+            if let egglog::ast::Expr::Call(f, xs) = e {
+                if let ("Id", [egglog::ast::Expr::Lit(egglog::ast::Literal::Int(int))]) =
+                    (f.as_str(), xs.as_slice())
+                {
+                    return Ok(Id(*int));
+                }
+            }
+            Err(String::from("invalid Id"))
+        }
+        fn egglog_expr_to_expr(e: &egglog::ast::Expr) -> Result<Expr, String> {
+            match e {
+                egglog::ast::Expr::Lit(_) => Err(String::from("literals must be wrapped")),
+                egglog::ast::Expr::Var(s) => Err(format!("unknown variable {:?}", s.as_str())),
```

Comment:
Maybe a more descriptive error would say "expected a grounded term, got a term with a variable in it: "